# Анализ базы данных сервиса для чтения книг

# Описание проекта

В наших руках крупный сервис для чтения книг по подписке

# Цель

Проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Описание данных

**Таблица `books` - данные о книгах**

- `book_id` — идентификатор книги
- `author_id` — идентификатор автора
- `title` — название книги
- `num_pages` — количество страниц
- `publication_date` — дата публикации книги
- `publisher_id` — идентификатор издателя

**Таблица `authors` - данные об авторах**

- `author_id` — идентификатор автора
- `author` — имя автора

**Таблица `publishers` - данные об издательствах**

- `publisher_id` — идентификатор издательства
- `publisher` — название издательства

**Таблица `ratings` - данные о пользовательских оценках книг**

- `rating_id` — идентификатор оценки
- `book_id` — идентификатор книги
- `username` — имя пользователя, оставившего оценку
- `rating` — оценка книги

**Таблица `reviews` - данные о пользовательских обзорах**

- `review_id` — идентификатор обзора
- `book_id` — идентификатор книги
- `username` — имя автора обзора
- `text` — текст обзора

# Задачи

 1. Посчитать, сколько книг вышло после 1 января 2000 года
 
 
 2. Для каждой книги посчитать количество обзоров и среднюю оценку
 
 
 3. Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры
 
 
 4. Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками
 
 
 5. Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок

## Подключение к базе данных

In [ ]:
import pandas as pd
from sqlalchemy import create_engine

from IPython.display import Image
from IPython.core.display import HTML 
# импортируем библиотеки

In [ ]:
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
# устанавливаем параметры

In [ ]:
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])

In [ ]:
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# сохраняем коннектор

## Изучение таблиц

In [ ]:
def result(query):
    display(pd.io.sql.read_sql(query, con = engine))
# функция для вывода 

### Таблица books

In [ ]:
books = ''' SELECT * 
            FROM books
        '''

In [ ]:
result(books)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

В таблице **books** представлена информация о 1000 книгах

### Таблица authors

In [ ]:
authors = ''' SELECT * 
              FROM authors
          '''

In [ ]:
result(authors)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

В таблице **authors** представлена информация о 636 авторах

### Таблица publishers

In [ ]:
publishers = ''' SELECT * 
                 FROM publishers
             '''

In [ ]:
result(publishers)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

В таблице **publishers** представлена информация о 340 издательствах

### Таблица ratings

In [ ]:
ratings = ''' SELECT * 
              FROM ratings
          '''

In [ ]:
result(ratings)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

Пользователи поставили 6456 оценок книгам

### Таблица reviews

In [ ]:
reviews = ''' SELECT * 
              FROM reviews
          '''

In [ ]:
result(reviews)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

Пользователями было оставлено 2793 обзора

## Решение задач

In [ ]:
Image(url="https://pictures.s3.yandex.net/resources/scheme_1589269096.png")
# посмотрим на схему базы данных

### Cколько книг вышло после 1 января 2000 года

In [ ]:
task_1 = ''' SELECT COUNT(book_id)
             FROM books
             WHERE publication_date > '2000-01-01' 
        '''

In [ ]:
result(task_1)

count
0    819

**Ответ:**

819 книг было выпущено после 1 января 2000 года

###  Количество обзоров и средняя оценка для каждой книги

In [ ]:
task_2 = ''' SELECT b.book_id, b.title, reviews_count, avg_rating
             FROM books AS b
             LEFT JOIN (
                        SELECT rev.book_id, COUNT(rev.review_id) AS reviews_count
                        FROM reviews AS rev
                        GROUP BY rev.book_id) AS tab_1 ON b.book_id = tab_1.book_id
             LEFT JOIN (
                        SELECT rat.book_id, ROUND(AVG(rat.rating),2) AS avg_rating
                        FROM ratings AS rat
                        GROUP BY rat.book_id) AS tab_2 ON b.book_id = tab_2.book_id         
         '''

In [ ]:
result(task_2)

book_id                                              title  \
0        652          The Body in the Library (Miss Marple  #3)   
1        273                                          Galápagos   
2         51                           A Tree Grows in Brooklyn   
3        951  Undaunted Courage: The Pioneering First Missio...   
4        839                                        The Prophet   
..       ...                                                ...   
995       64                                Alice in Wonderland   
996       55          A Woman of Substance (Emma Harte Saga #1)   
997      148                                          Christine   
998      790  The Magicians' Guild (Black Magician Trilogy  #1)   
999      828                           The Plot Against America   

     reviews_count  avg_rating  
0              2.0        4.50  
1              2.0        4.50  
2              5.0        4.25  
3              2.0        4.00  
4              4.0        4.29  
..             ...         ...  
995            4.0        4.23  
996            2.0        5.00  
997            3.0        3.43  
998            2.0        3.50  
999            2.0        3.00  

[1000 rows x 4 columns]

**Ответ:**
    
Таблица с количеством обзоров и средней оценкой для каждой книги    

### Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [ ]:
task_3 = ''' SELECT p.publisher, COUNT(b.book_id) AS books_count
             FROM books AS b
             LEFT JOIN publishers AS p ON b.publisher_id = p.publisher_id
             WHERE b.num_pages > 50
             GROUP BY p.publisher
             ORDER BY books_count DESC
             LIMIT 1
        '''

In [ ]:
result(task_3)

publisher  books_count
0  Penguin Books           42

**Ответ:**

Издательство Penguin Books выпустило 42 книги, толще 50 страниц

### Автор с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками

In [ ]:
task_4 = ''' SELECT a.author, ROUND(AVG(r.rating), 2) AS avg_rating
             FROM books AS b
             LEFT JOIN authors AS a ON b.author_id = a.author_id 
             LEFT JOIN ratings AS r ON b.book_id = r.book_id
             WHERE b.book_id IN (
                                 SELECT book_id
                                 FROM ratings 
                                 GROUP BY book_id
                                 HAVING COUNT(rating_id) > 50)
             GROUP BY a.author
             ORDER BY AVG(r.rating) DESC
             LIMIT 1
         '''

In [ ]:
result(task_4)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

**Ответ:**

J.K. Rowling - автор с самой высокой средней оценкой книг (4.29)

### Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [ ]:
task_5 = '''  SELECT ROUND(AVG(count_rev)) AS avg_reviews_count
              FROM (
                      SELECT username, COUNT(review_id) AS count_rev
                      FROM reviews
                      WHERE username IN (
                                          SELECT username
                                          FROM ratings 
                                          GROUP BY username
                                          HAVING COUNT(rating_id) > 50)
                     GROUP BY username) AS tab_1 
        '''

In [ ]:
result(task_5)

avg_reviews_count
0               24.0

**Ответ:**

Пользователи, которые поставили больше 50 оценок, в среднем пишут 24 отзыва

## Вывод:

Проанализировали базу дааных сервиса для чтения книг и узнали: 

- 819 книг было выпущено после 1 января 2000 года
- Издательство Penguin Books выпустило 42 книги, толще 50 страниц
- J.K. Rowling - автор с самой высокой средней оценкой книг (4.29)
- Пользователи, которые поставили больше 50 оценок, в среднем пишут 24 отзыва